### DISCUSSION TITLE
Running on headless sever

### DISCUSSION THREAD

**Discussion by TetsujinXLIV** [fat-tire/resolve#55] [Q&A]
Just found this project. I was wondering if it would be possible to run this on my Ubuntu Server that doesn't have a GUI. Can you have the application window open in a browser? Or does it require running this on something like Ubuntu Desktop?

Thanks!

**Comment by fat-tire** [comment]
Great question-- if you are using Android Studio, I think you should be able to run the container using:

`./studio.sh /bin/bash`

At that point, you can use resolve without a GUI-- such as via a [python or LUA script.](https://deric.github.io/DaVinciResolve-API-Docs/) .

If, on the other hand, you're asking if resolve can be run WITH an GUI but without that GUI appearing on the original device-- yes, but I'm not 100% sure the best way for the best performance.  There are options like X11 forwarding via SSH as well as remote desktop sharing via [Ubuntu Screen sharing](https://virola.io/articles/how-to-set-up-screen-sharing-on-linux-ubuntu). 

Basically you'd go to **Settings**->**System**->**Desktop Sharing** and enable **Remote Contro**l and then use a remote desktop app with RDP on the other machine to connect to the host and access the container that way.

If you wanted to connect directly to the container and bypass the host, you could install an RDP server on the container and access that way.  I should note there is also [VNC](https://help.realvnc.com/hc/en-us/articles/360003474752-How-do-I-use-VNC-Server-in-Virtual-Mode-on-Linux-in-conjunction-with-the-system-Xorg-server) and other methods.

For fun, here is just a guess of what would be done manually to make this work via [x11 forwarding with SSH](https://unix.stackexchange.com/questions/12755/how-to-forward-x-over-ssh-to-run-graphics-applications-remotely).

1. [install sshd in the container](https://linuxconfig.org/how-to-install-start-and-connect-to-ssh-server-on-fedora-linux) by `./studio.sh /bin/bash` and then something like `dnf install openssh-server`.  (Really you'll want to put this process into the `Dockerfile` and rebuild so that it's there every time you run the container)
2. Configure `/etc/ssh/sshd_config` to choose your port and any other settings you want, and to make sure that x11 forwarding is enabled. (also put in the `Dockerfile`)
3. You will need to adjust `studio.sh` to open port 2222 (or whatever) from the container to the host something like `--port 2222:2222` or whatever.  The Ip should probably be `0.0.0.0` to make it accessible from other machines.  I'm using port 2222 rather than 22 so it doesn't conflict with sshd already running on the host and using a port number that won't require root if you're using podman instead of docker.
4. Then you'd set up your ssh pubkeys or password to access the machine running resolve.
5. Now run the container and try to connect from the other machine using X11 forwarding.  
6. If that works, you can hopefully run `/opt/resolve/bin/resolve` and have it pop up on your remote machine.  (I'm assuming this is a linux machine running x11)
7. As for sound, I think pulseaudio can [forward sound](https://askubuntu.com/questions/371687/how-to-carry-audio-over-ssh) as can pipewire.  See [here](https://superuser.com/questions/231920/forwarding-audio-like-x-in-ssh) for more.  It might "just work" I'm not really sure.

I'd be very curious if you can get this working.  You can also investigate  [XRDP](https://www.xrdp.org/) which should automatically work with sound, clipboard transfer, etc.

I don't see any reason why you couldn't do it though.  The only potential technical issue I can think of would be latency-- especially if you're streaming giant videos across long distances with lower bandwidth connections.

Resolve, incidentally, has a cloud feature that lets multiple people collaborate simultaneously on the same project, a bit like Google Docs.  Each side should ideally have access to some version of the media, although a lower quality proxy resolution can be substituted for the full version.

I'm not aware of a X11->Web Server gateway, although perhaps there is one (?)
